src/liballoc/collections/binary_heap.rs

```diff
@@ -899,6 +920,47 @@ impl<T> BinaryHeap<T> {
         Drain { iter: self.data.drain(..) }
     }
 
+    /// Returns an iterator which retrieves elements in heap order.
+    /// The retrieved elements will be removed from the original heap.
+    ///
+    /// Note:
+    /// * Unlike other `.drain()` methods, this method removes elements *lazily*.
```

Comment:
Having `drain_sorted` behave differently than other `drain` methods with respect to optimistically clearing capacity seems unfortunate. Combined with the inconsistent order of any remaining elements in `drop` it's looking like a method that isn't going to behave the way you expect unless you hold it a specific way.

I feel like we'll need a few examples that are at least close to some real-world usage to get a sense of whether these trade-offs really are reasonable.

Reply:
In 30e8f65, I made `.drain_sorted()` to always remove elements in heap order.
This simplified doc example a lot.